### PR TITLE
Fully working merge pipeline 

### PR DESCRIPTION
**Pull request by piyushrpt**
* Opening PR in draft mode to leave this here
* Will keep making minor changes to it as we gather feedback
* Tested on the provided sample dataset
* Builds on #11 and #13

### CODE REVIEW

**Comment by scottstanie** [comment]
(I realize that you're actively pushing, so if you already noticed, please disregard) I saw the second to last commit has `from Pathlib import Path`, which is failing from the capital C.

Have you run into intermittent test failures for the `unwrap_one` test?

If I do 
```
$ pytest --randomly-seed=2713237683 test/mcf/test_ortools.py::test_unwrap_one
```

I get 
```
FAILED test/mcf/test_ortools.py::test_unwrap_one - AssertionError: assert 2.6645352591003757e-15 > 1
```

but other seeds work okay for me.

**Comment by piyushrpt** [comment]
😢 - good catch. I will include it as part of the next couple of commits later today.

**Comment by scottstanie** [comment]
Have you come across this ORTools error? 

```
$ python -m spurt.workflows.emcf -i . -w 3
2024-05-21 17:45:16,337 [49229] [INFO] spurt: Using Hop3 Graph in time with 21 epochs.
2024-05-21 17:45:16,338 [49229] [INFO] spurt: Using existing tiles file: emcf_tmp/tiles.json
2024-05-21 17:45:16,338 [49229] [INFO] spurt: Processing tile: 1
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Time steps: 57
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Number of points: 421201
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Temporal: Number of interferograms: 57
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Temporal: Number of links: 1262032
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Temporal: Number of cycles: 37
2024-05-21 17:45:28,906 [49229] [INFO] spurt: Temporal: Preparing batch 1/26
2024-05-21 17:45:29,033 [49229] [INFO] spurt: Temporal: Unwrapping batch 1/26
Processing batch of 50000 with 3 threads
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/staniewi/miniconda3/envs/mapping-311/lib/python3.11/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
                    ^^^^^^^^^^^^^^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/mcf/_ortools.py", line 360, in wrap_solve_mcf
    return (ind, solve_mcf(es, ed, rr, cc, rc))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/mcf/_ortools.py", line 323, in solve_mcf
    raise ValueError(errmsg)
ValueError: MCF unbalanced. Sum of residues in non-zero.
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/staniewi/repos/spurt/src/spurt/workflows/emcf/__main__.py", line 5, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/workflows/emcf/_cli.py", line 96, in main
    unwrap_tiles(stack, g_time, gen_settings, slv_settings)
  File "/Users/staniewi/repos/spurt/src/spurt/workflows/emcf/_unwrap.py", line 54, in unwrap_tiles
    uw_data = solver.unwrap_cube(wrap_data)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/workflows/emcf/_solver.py", line 132, in unwrap_cube
    grad_space: np.ndarray = self.unwrap_gradients_in_time(
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/workflows/emcf/_solver.py", line 223, in unwrap_gradients_in_time
    flows = self._solver_time.residues_to_flows_many(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/staniewi/repos/spurt/src/spurt/mcf/_ortools.py", line 277, in residues_to_flows_many
    for res in mp_tasks:
  File "/Users/staniewi/miniconda3/envs/mapping-311/lib/python3.11/multiprocessing/pool.py", line 873, in next
    raise value
ValueError: MCF unbalanced. Sum of residues in non-zero.
```

File: src/spurt/workflows/emcf/_unwrap.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        # integer cycle shifts between tiles.
+        ifgs = g_time.links
+        phase_offset = spurt.mcf.utils.phase_diff(
+            wrap_data.data[ifgs[0, :], 0], wrap_data.data[ifgs[1, :], 0]
+        )
+
+        _dump_tile_to_h5(tfname, uw_data, phase_offset, g_space, tiledata["tiles"][tt])
+        logger.info(f"Wrote tile {tt + 1} to {tfname}")
+        wrap_data = None
+        uw_data = None
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_unwrap.py]: Done

**Comment by piyushrpt** [comment]
Hmm .. I haven't run into that unbalanced error before. 
This should ensure that sum is zero - https://github.com/piyushrpt/spurt/blob/f75d2b39b3680b242e2298b11df7367d0976fdd1/src/spurt/workflows/emcf/_solver.py#L219

I will add a cast to integer here just like in mcf solver.

**Comment by piyushrpt** [comment]
The other possibility is that there are nans in the data. In which case, the sum of residues will end up being a nan and that breaks the balanced check

**Comment by scottstanie** [comment]
You were correct, there were a few nan pixels in a couple images.

Perhaps we could build in some nan check for a clearer error, or a conversion like `nan_to_num` if there's no harm in converting to zero.

@gmgunter thoughts on those options?

**Comment by piyushrpt** [comment]
There are only a couple of places where pixel selection takes place. Example:
https://github.com/piyushrpt/spurt/blob/911f0d4021dd7ee0488017ccddcf56ced997414e/src/spurt/workflows/emcf/_tiling.py#L21

I would suggest not needing nan checks on phase data - but if temporal coherence can be set to nan / zero for such pixels - the workflow can already handle it.

**Comment by scottstanie** [comment]
Just noting that my tests are running through, while there's small possible changes/additions, the biggest thing that will help calling this from `dolphin` will be to split the `_cli.main` into one function that gets args from `sys.argv` (something like `_get_cli_args(args: list[str] = None)`, and another that has all the inputs listed out that's more callable in the middle of a workflow.

Since that might take a little code juggling, we can wait until this is merged to split that up

**Comment by piyushrpt** [comment]
This has been rebased and ready for review. I will wait for this to be merged before starting the next set.

File: src/spurt/graph/utils.py [hunk 2]
```diff
@@ -21,3 +31,18 @@ def order_points(p: tuple[int, int]) -> tuple[int, int]:
...
+    links = graph.links
+    nlinks = len(links)
+    data = np.ones(2 * nlinks, dtype=int)
+    data[0::2] = -1
+    amat = csr_matrix(
+        (
+            data,
+            (np.arange(2 * nlinks, dtype=int) // 2, links.flatten()),
+        )
+    )
```
Comment (scottstanie) [inline src/spurt/graph/utils.py:47]: how about this to give more context for what `amat` is?

```suggestion
    # Build the incidence matrix of the graph:
    # Columns represent nodes, rows represent links (edges)
    # Each row has one -1 and one +1 for the source/dest. node index
    row_indices = np.arange(2 * nlinks, dtype=int) // 2
    col_indices = links.flatten()
    amat = csr_matrix((data, (row_indices, col_indices)))
```

**Comment by scottstanie** [comment]
```python
from ._settings import GeneralSettings, MergerSettings, SolverSettings, TilerSettings
from ._solver import EMCFSolver as Solver

__all__ = [
    "Solver",
    "MergerSettings",
    "GeneralSettings",
    "SolverSettings",
    "TilerSettings",
]
```

`GeneralSettings` [currently misspelled in the init](https://github.com/piyushrpt/spurt/blob/feat/dirichlet/src/spurt/workflows/emcf/__init__.py#L6), and having the other options exposed seems useful

File: src/spurt/workflows/emcf/_output.py [hunk 3]
```diff
@@ -143,37 +161,62 @@ def write_single_tile(
...
+        assert tile.correction_level == 1, "Only one offset supported"
 
     # Create full sized array
+    model = np.zeros(shape, dtype=np.float32)
+    cnt = np.zeros(shape, dtype=np.int16)
+    minval = np.full(shape, np.inf, dtype=np.float32)
+    maxval = np.full(shape, -np.inf, dtype=np.float32)
+
+    # Now iterate over tiles and compute average
+    for _, tile in tiles.items():
```
Comment (gmgunter) [inline src/spurt/workflows/emcf/_output.py:188]:
```suggestion
    for tile in tiles.values():
```

File: src/spurt/workflows/emcf/_output.py [hunk 4]
```diff
@@ -143,37 +161,62 @@ def write_single_tile(
...
+        coords = tile.coords.astype(np.int32)
+        c0 = coords[:, 0]
+        c1 = coords[:, 1]
+
+        data = tile.uw_phase
+        minval[c0, c1] = np.minimum(data, minval[c0, c1])
+        maxval[c0, c1] = np.maximum(data, maxval[c0, c1])
+
+        model[coords[:, 0], coords[:, 1]] += tile.uw_phase
+        cnt[coords[:, 0], coords[:, 1]] += 1
```
Comment (gmgunter) [inline src/spurt/workflows/emcf/_output.py:198]:
```suggestion
        model[c0, c1] += tile.uw_phase
        cnt[c0, c1] += 1
```

File: src/spurt/workflows/emcf/_output.py [hunk 5]
```diff
@@ -143,37 +161,62 @@ def write_single_tile(
...
+    mask = cnt != 0
+    model[~mask] = np.nan
+    model[mask] /= cnt[mask]
+    diff = maxval - minval
+    diff[cnt < 2] = np.nan
+    diff[np.isinf(diff)] = np.nan
+
+    # Now iterate over tiles and wrap around model
     arr = np.full(shape, np.nan, dtype=np.float32)
+    for _, tile in tiles.items():
```
Comment (gmgunter) [inline src/spurt/workflows/emcf/_output.py:209]:
```suggestion
    for tile in tiles.values():
```

File: test/utils/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+import numpy as np
+
+import spurt
+
+
+def test_cpu_count():
+    ncpu = spurt.utils.get_cpu_count()
+
+    # Check that atleast one is returned
```
Comment (gmgunter) [inline test/utils/test_utils.py:11]:
```suggestion
    # Check that at least one is returned
```

**Review by gmgunter — APPROVED** [review]
LGTM only minor comments

File: test/utils/test_dirichlet.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+def test_dirichlet():
+
+    # Create some random points and per-point data.
+    pts = np.random.randn(4000).reshape((2000, 2))
+
+    cutoff = -1
+
+    # Choose points with an x-value below cutoff to be the fixed (boundary)
+    # values. Choose 1 for poisitive y-values and -1 for negative y-values.
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply test/utils/test_dirichlet.py:16]: Done

File: src/spurt/graph/utils.py (hunk 2 above)
Comment (piyushrpt) [reply src/spurt/graph/utils.py:47]: Done

File: src/spurt/graph/utils.py [hunk 8]
```diff
@@ -21,3 +31,18 @@ def order_points(p: tuple[int, int]) -> tuple[int, int]:
     if p[0] <= p[1]:
         return (p[0], p[1])
     return (p[1], p[0])
+
+
+def graph_laplacian(graph: GraphInterface) -> csc_matrix:
+    """Compute the graph laplacian."""
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/graph/utils.py:37]: Done

File: src/spurt/mcf/utils.py [hunk 9]
```diff
@@ -38,7 +38,9 @@ def sign_nonzero(x: float) -> int:
     return -1
 
 
-def flood_fill(indata: np.ndarray, links: np.ndarray, flows: np.ndarray):
+def flood_fill(
+    indata: np.ndarray, links: np.ndarray, flows: np.ndarray, mode: str | None = None
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/utils.py:42]: Done

File: src/spurt/mcf/utils.py [hunk 10]
```diff
@@ -58,23 +60,44 @@ def flood_fill(indata: np.ndarray, links: np.ndarray, flows: np.ndarray):
...
+        errmsg = "mode must be set to points or gradients."
+        raise ValueError(errmsg)
+
     if len(links) != len(flows):
         errmsg = f"Dimension mismatch - {links.shape} vs {flows.shape}"
         raise ValueError(errmsg)
 
-    if len(indata) == len(links):
+    if mode not in ["points", "gradients"]:
+        errmsg = f"mode can be points or gradients. got {mode}"
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/utils.py:80]: Done

File: src/spurt/utils/merge.py [hunk 11]
```diff
@@ -232,12 +223,9 @@ def cb(*_):
     # 1 and 2 above.
     rn = rhs - np.dot(mat, x) if use_normal_eqs else None
 
-    if logger is not None:
+    if enable_logging is not None:
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/merge.py:226]: Done

File: src/spurt/mcf/utils.py [hunk 12]
```diff
@@ -58,23 +60,44 @@ def flood_fill(indata: np.ndarray, links: np.ndarray, flows: np.ndarray):
...
+        Can be one of 'points' or 'gradients'
 
     Returns
     -------
     unwrapped : np.ndarray
         Unwrapped phase in radians. Same size as indata.
     """
+    if mode is None:
+        errmsg = "mode must be set to points or gradients."
+        raise ValueError(errmsg)
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/mcf/utils.py:73]: Done

File: test/utils/test_utils.py (hunk 6 above)
Comment (piyushrpt) [reply test/utils/test_utils.py:11]: Done

File: src/spurt/workflows/emcf/_output.py (hunk 5 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_output.py:209]: Done

File: src/spurt/workflows/emcf/_output.py (hunk 4 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_output.py:198]: Done

File: src/spurt/workflows/emcf/_output.py (hunk 3 above)
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_output.py:188]: Done

File: src/spurt/workflows/emcf/_output.py [hunk 13]
```diff
@@ -143,37 +161,62 @@ def write_single_tile(
...
-    fname = gen_settings.unw_filename(ifg[0], ifg[1])
     if fname.is_file():
         logger.info(
             f"{fname!s} for band {idx + 1} already exists. Skipping writing ..."
         )
         return
 
     # check that we are looking at the same band in all tiles
-    for tile in tiles:
+    for _, tile in tiles.items():
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_output.py:177]: Done

File: src/spurt/utils/merge.py [hunk 14]
```diff
@@ -147,11 +152,11 @@ def l2_min_cg(
         Sparse real system to solve.
     b : vector-like
         Right-hand side.
-    logger : logger-like
-        Optional, logger to which diagnostic messages will be sent.
+    enable_logging : bool
+        Optional, Enable logger to capture diagnostic messages.
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/merge.py:156]: The level of logging messages are not very verbose here. They are meant to be used at INFO level only.

File: src/spurt/workflows/emcf/_settings.py [hunk 15]
```diff
@@ -114,11 +116,39 @@ class TilerSettings:
...
 
+    target_points_per_tile: int = 800000
     max_tiles: int = 16
     target_points_for_generation: int = 120000
-    target_points_per_tile: int = 800000
+    dilation_factor: float = 0.05
+
+    def __post_init__(self):
+        if self.max_tiles < 1:
+            errmsg = f"max_tiles must be atleast 1, got {self.max_tiles}"
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_settings.py:136]: Done

File: src/spurt/workflows/emcf/_tiling.py [hunk 16]
```diff
@@ -15,91 +12,52 @@ def get_tiles(
     gen_settings: GeneralSettings,
     tile_settings: TilerSettings,
 ) -> None:
-    """Get the tiles for given stack."""
+    """Generate tiles based on settings.
+
+    Create a json file with tile information in the intermediate folder.
+    Tiles are not regenerated if a json file is already present.
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_tiling.py:18]: Done

File: src/spurt/utils/_tiling.py [hunk 17]
```diff
@@ -1,237 +1,251 @@
...
-            for jj in range(tiles_per_dim):
-                cdiff = cols[jj + 1] - cols[jj]
-                c0 = int(max(0, cols[jj] - self._dilation * cdiff))
-                c1 = int(min(self._shape[1], cols[jj + 1] + self._dilation * cdiff))
-                tiles.append([r0, c0, r1, c1])
-
-        # Track neighbors
-        nbrs = []
-        ntiles = len(tiles)
+        return TileSet(shape, tiles)
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/_tiling.py:100]: Done

File: src/spurt/utils/_tiling.py [hunk 18]
```diff
@@ -1,237 +1,251 @@
...
+            "shape": self.shape,
+            "tiles": tiles,
+        }
+        with fname.open(mode="w") as fid:
+            fid.write(json.dumps(jdict, indent=4))
+
+    @classmethod
+    def single_tile(cls, shape: tuple[int, int]) -> TileSet:
+        """Return tileset with single tile corresponding to shape."""
+        return TileSet(shape, [BBox.from_shapely_bounds((0, 0, shape[0], shape[1]))])
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/_tiling.py:118]: Done

File: src/spurt/utils/_tiling.py [hunk 19]
```diff
@@ -1,237 +1,251 @@
...
+    def ntiles(self) -> int:
+        return len(self._tiles)
 
-        # If only 1 tile was requested
-        if tiles_per_dim == 1:
-            self._tiles = np.array([[0, 0, self._shape[0], self._shape[1]]])
-            return
+    @classmethod
+    def from_json(cls, fname: Path) -> TileSet:
+        """Load tiles from a json file."""
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/_tiling.py:91]: Done

File: src/spurt/utils/_tiling.py [hunk 20]
```diff
@@ -1,237 +1,251 @@
...
-                c1 = int(min(self._shape[1], cols[jj + 1] + self._dilation * cdiff))
-                tiles.append([r0, c0, r1, c1])
-
-        # Track neighbors
-        nbrs = []
-        ntiles = len(tiles)
+        return TileSet(shape, tiles)
+
+    def to_json(self, fname: Path) -> None:
+        """Write tiles to a json file."""
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/_tiling.py:103]: Done

File: src/spurt/utils/_tiling.py [hunk 21]
```diff
@@ -1,237 +1,251 @@
...
+            rdiff = tt.xmax - tt.xmin
+            r0 = int(max(0, tt.xmin - factor * rdiff))
+            r1 = int(min(shape[0], tt.xmax + factor * rdiff))
+            cdiff = tt.ymax - tt.ymin
+            c0 = int(max(0, tt.ymin - factor * cdiff))
+            c1 = int(min(shape[1], tt.ymax + factor * cdiff))
 
-    This is based on the nice implementation found here:
-    https://mathoverflow.net/questions/412127/partitioning-unit-square-with-equal-frequency-rectangles
+            tiles.append(BBox.from_shapely_bounds([r0, c0, r1, c1]))
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/utils/_tiling.py:147]: Done

File: src/spurt/workflows/emcf/_merge.py [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+        fnames.append(gen_settings.unw_filename(dates[ifg[0]], dates[ifg[1]]))
+
+    # If we only have to write one tile
+    # Nothing to merge - just write to geotiff
+    if len(tiles) == 1:
+        logger.info(f"Writing single tile output to {gen_settings.output_folder}")
+        write_single_tile(tiles[0], fnames, tiledata.shape)
+        return
+
+    # Create overmap lap for the graph
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_merge.py:48]: Done

File: src/spurt/workflows/emcf/_merge.py [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+                    continue
+                idx_i, idx_j = _get_common_overlap(gen_settings, ii, jj)
+                diff = tile_i.uw_phase[idx_i] - tiles[jj].uw_phase[idx_j]
+                logger.info(f"Overlap between {ii}, {jj} : {np.max(np.abs(diff))}")
+
+
+def _get_overlap_map(
+    gen_settings: GeneralSettings,
+) -> dict[int, list[int]]:
+    """Return overlap map for each tile."""
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_merge.py:152]: Done

File: src/spurt/workflows/emcf/_merge.py [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+        # Write file to output
+        write_merged_band(tiles, fname, ii, tiledata.shape)
+
+
+def _adjust_tiles(
+    tiles: dict[int, Tile],
+    overlap_map: dict[int, list[int]],
+    gen_settings: GeneralSettings,
+    max_degree: int,
+    debug_stats: bool | None = None,
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_merge.py:84]: Done

File: src/spurt/workflows/emcf/_merge.py [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+    """Dirichlet-based tile adjustment described in [1]_.
+
+    References
+    ----------
+    .. [1] M. T. Calef, Olsen K. M., & Agram P. S., "Merging Point Data for
+           InSAR Deformation Processing", in arXiv preprints arXiv:2405.06838, 2024.
+    """
+    if debug_stats is None:
+        debug_stats = False
+
```
[the remark this answers is not in the document]
Comment (piyushrpt) [reply src/spurt/workflows/emcf/_merge.py:95]: Done

**Pull request merged**